File: lamquant-firmware/src/neural/fsq.rs
//! FSQ (Finite Scalar Quantization) — quantize latent dimensions to a
//! discrete grid, then encode 4-tuple as a single flat index.
//!
//! Adaptive scaling: tracks rolling RMS of the input EEG and adjusts the
//! Q31 projection scale to prevent dead codes (weak signal) and clipping
//! (strong signal). The "50" threshold matches the silicon shackle clamp
//! range (±50 in signal units).

use core::sync::atomic::{AtomicI32, Ordering};

/// Default Q31 projection scale. Wired from the trained model via
/// `firmware_export.py` in Phase 5; for now use a placeholder.
const DEFAULT_FSQ_QUANT_SCALE_Q31: i32 = 0x4000_0000; // 0.5 in Q31

/// FSQ levels per dimension (Gen 7.7: production preset).
/// Indexes into a 4D lattice. Phase 5 will expose runtime preset switching.
pub const FSQ_LEVELS: [i32; 4] = [8, 16, 16, 32];

/// Adaptive scale, rebuilt every window from RMS.
static ADAPTIVE_FSQ_SCALE: AtomicI32 = AtomicI32::new(DEFAULT_FSQ_QUANT_SCALE_Q31);

/// Q31 multiply: `(a * b) >> 31`, signed.
#[inline(always)]
fn mul_q31(a: i32, b: i32) -> i32 {
    ((a as i64 * b as i64) >> 31) as i32
}

/// Integer square root via shift-and-subtract. O(16) iterations for u32.
/// Returns `floor(sqrt(x))`.
fn isqrt32(mut x: u32) -> u32 {
    if x == 0 {
        return 0;
    }
    let mut result = 0u32;
    let mut bit = 1u32 << 30; // highest power of 4

    while bit > x {
        bit >>= 2;
    }
    while bit != 0 {
        if x >= result + bit {
            x -= result + bit;
            result = (result >> 1) + bit;
        } else {
            result >>= 1;
        }
        bit >>= 2;
    }
    result
}

/// Update adaptive FSQ gain from rolling RMS of the current window.
///
/// Called once per window with the Q31 ADC buffer. Adjusts the projection
/// scale so weak signals fill bins and strong signals don't clip.
pub fn update_adaptive_gain(eeg_buffer: &[i32]) {
    if eeg_buffer.is_empty() {
        return;
    }

    let mut sum_sq: i64 = 0;
    for &val in eeg_buffer {
        sum_sq += (val as i64) * (val as i64);
    }

    let mean_sq = (sum_sq / eeg_buffer.len() as i64) as u32;
    let rms = isqrt32(mean_sq);

    let new_scale = if rms > 50 {
        DEFAULT_FSQ_QUANT_SCALE_Q31 / ((rms / 50) + 1) as i32
    } else {
        DEFAULT_FSQ_QUANT_SCALE_Q31
    };

    ADAPTIVE_FSQ_SCALE.store(new_scale, Ordering::Relaxed);
}

/// Quantize a single Q31 activation to the nearest FSQ grid point.
///
/// Maps continuous Q31 value to integer in `[-levels/2, +levels/2]`.
#[inline]
fn quantize_scalar(val: i32, levels: i32) -> i32 {
    let half_bound = levels >> 1;
    let scale = ADAPTIVE_FSQ_SCALE.load(Ordering::Relaxed);
    let grid_estimate = mul_q31(val, scale);
    grid_estimate.clamp(-half_bound, half_bound)
}

/// Run FSQ on a 4-element activation vector. Returns the flat lattice index.
///
/// Mixed-radix encoding: index = Σᵢ (q_val[i] · stride[i])
/// where stride[i] = ∏_{j<i} FSQ_LEVELS[j].
///
/// The flat index uniquely identifies one point in the 4D FSQ lattice and
/// is the input to the entropy coder (rANS context-adaptive table lookup).
pub fn run_translation(network_activations_4d: &[i32; 4]) -> u32 {
    let mut single_index: u32 = 0;
    let mut implicit_stride: u32 = 1;

    for i in 0..4 {
        let q_val = quantize_scalar(network_activations_4d[i], FSQ_LEVELS[i]);
        // Shift from [-half, +half] to [0, levels-1].
        let base_index = q_val + (FSQ_LEVELS[i] >> 1);
        single_index += (base_index as u32).wrapping_mul(implicit_stride);
        implicit_stride = implicit_stride.wrapping_mul(FSQ_LEVELS[i] as u32);
    }
    single_index
}

#[cfg(all(test, feature = "host-verify"))]
mod tests {
    use super::*;

    #[test]
    fn isqrt32_basics() {
        assert_eq!(isqrt32(0), 0);
        assert_eq!(isqrt32(1), 1);
        assert_eq!(isqrt32(4), 2);
        assert_eq!(isqrt32(9), 3);
        assert_eq!(isqrt32(15), 3); // floor
        assert_eq!(isqrt32(100), 10);
        assert_eq!(isqrt32(10000), 100);
    }

    #[test]
    fn quantize_clamps_to_grid_bounds() {
        // Max activation should clamp to +levels/2.
        let q = quantize_scalar(i32::MAX, 8);
        assert!(q.abs() <= 4, "out of grid range: {q}");
    }

    #[test]
    fn translation_returns_index_in_lattice_size() {
        let lattice_size: u32 = FSQ_LEVELS.iter().product::<i32>() as u32;
        let zeros = [0i32; 4];
        let idx = run_translation(&zeros);
        assert!(idx < lattice_size, "FSQ index {idx} >= {lattice_size}");
    }
}

File: reference_implementations/c_firmware/neural/fsq.c
#include <stdint.h>
#include <stddef.h>
#include "../firmware_export/fsq_lattice.h"
#include "../firmware_export/focal_net_weights.h"
#include "../core/math_utils.h"
#include "fsq.h"

// Adaptive FSQ scaling: adjusts quantization grid based on rolling signal RMS.
// Ensures weak signals fill the FSQ bins (preventing dead codes) and strong
// signals don't clip.
static volatile int32_t adaptive_fsq_scale = FSQ_QUANT_SCALE_Q31;

/*
 * Integer square root (no float, no libm).
 * Uses binary search: O(16) iterations for 32-bit input.
 * Returns floor(sqrt(x)).
 */
static uint32_t isqrt32(uint32_t x) {
    if (x == 0) return 0;
    uint32_t result = 0;
    uint32_t bit = 1u << 30;  // Start from highest power of 4

    while (bit > x) bit >>= 2;
    while (bit != 0) {
        if (x >= result + bit) {
            x -= result + bit;
            result = (result >> 1) + bit;
        } else {
            result >>= 1;
        }
        bit >>= 2;
    }
    return result;
}

/*
 * Update adaptive FSQ gain from a rolling RMS estimate.
 *
 * Called once per window with the current ADC buffer (Q31 samples).
 * Computes integer RMS and adjusts the FSQ projection scale:
 *   - High RMS (>50 in signal units): compress scale to avoid clipping
 *   - Low/nominal RMS: use default FSQ_QUANT_SCALE_Q31
 *
 * The "50" threshold corresponds to the silicon shackle clamp range (±50).
 * Signals near the clamp boundary need reduced scale to avoid saturation
 * in the FSQ lattice.
 */
void fsq_update_adaptive_gain(const int32_t* eeg_buffer, int len) {
    if (len <= 0) return;

    int64_t sum_sq = 0;
    for (int i = 0; i < len; i++) {
        int32_t val = eeg_buffer[i];
        sum_sq += (int64_t)val * val;
    }

    // Integer RMS: isqrt(sum_sq / len)
    uint32_t mean_sq = (uint32_t)(sum_sq / (int64_t)len);
    uint32_t rms = isqrt32(mean_sq);

    // Adaptive scaling: compress when signal is near the ±50 shackle boundary.
    // The divisor (rms/50 + 1) is always >= 1, preventing division by zero
    // and ensuring scale never exceeds the nominal value.
    if (rms > 50) {
        adaptive_fsq_scale = FSQ_QUANT_SCALE_Q31 / (int32_t)((rms / 50) + 1);
    } else {
        adaptive_fsq_scale = FSQ_QUANT_SCALE_Q31;
    }
}

/*
 * Quantize a single Q31 activation to the nearest FSQ grid point.
 *
 * Input range: Q31 scaled neural activations (output of ternary MAC).
 * The adaptive_fsq_scale (Q31 format) projects the activation onto the
 * discrete FSQ grid [-levels/2, +levels/2].
 *
 * mul_q31(val, scale) performs: (int32_t)((int64_t)val * scale >> 31)
 * This maps the continuous Q31 value to an integer grid index.
 */
static inline int32_t fsq_quantize_scalar(int32_t val, int32_t levels) {
    int32_t half_bound = levels >> 1;

    // Project activation onto FSQ grid using adaptive Q31 scale
    int32_t grid_estimate = mul_q31(val, adaptive_fsq_scale);

    // Clamp to valid grid range
    if (grid_estimate > half_bound)  grid_estimate = half_bound;
    if (grid_estimate < -half_bound) grid_estimate = -half_bound;

    return grid_estimate;
}

/*
 * Run FSQ on a 4-element activation vector.
 *
 * Converts 4 quantized grid values into a single flat index using
 * mixed-radix encoding: index = sum(q_val[i] * stride[i])
 * where stride[i] = product(FSQ_LEVELS[0..i-1]).
 *
 * The flat index uniquely identifies one point in the 4D FSQ lattice
 * and is the input to the entropy coder.
 */
uint32_t run_fsq_translation(int32_t* network_activations_4d) {
    uint32_t single_index = 0;
    uint32_t implicit_stride = 1;

    for (int i = 0; i < 4; i++) {
        int32_t q_val = fsq_quantize_scalar(network_activations_4d[i], FSQ_LEVELS[i]);
        // Shift from [-half, +half] to [0, levels-1] for unsigned index
        int32_t base_index = q_val + (FSQ_LEVELS[i] >> 1);
        single_index += (uint32_t)base_index * implicit_stride;
        implicit_stride *= (uint32_t)FSQ_LEVELS[i];
    }
    return single_index;
}
